src/api/syscall.c

```diff
@@ -55,14 +55,10 @@ exception_t handleInterruptEntry(void)
         handleSpuriousIRQ();
     }
 
-#ifdef CONFIG_KERNEL_MCS
     if (SMP_TERNARY(clh_is_self_in_queue(), 1)) {
-#endif
         schedule();
         activateThread();
```

Comment:
`IpiRemoteCall_VMCheckBoundNotification` on x86 looks like it is relying on the call to schedule on the way out of the ipi.  So there's potentially impact there if this were to be changed (and maybe it's broken on MCS as it is and could explain some MCS + SMP odd behavior reported in the past: https://github.com/seL4/seL4/pull/118#issuecomment-1473072760).

Reply:
I suppose the performance penalty for a `IpiRemoteCall_VMCheckBoundNotification` + `irq_reschedule_ipi` would be too big?

The MCS + SMP case here looks like it's incorrect and presents a race condition, as the target core's scheduling queues become a critical section.

Reply:
@andybui01 can you expand upon this a bit? I'm actually investigating a problem where a VM is getting starved of execution time, and as far as I can tell, the userspace VCPU execution loops aren't the culprit.

Reply:
I'm not really familiar with the x86 virtualization support but here's what I recall: currently the `VMCheckBoundNotification()` function that runs on the destination core may ask for a switch to a new thread. This new thread gets scheduled and activated in `handleInterruptEntry@syscall.c`. This can work because remote call IPIs operate with the acquired BKL of the calling core.

If we follow this PR so that `schedule()` checks that the _current_ core holds the BKL, this IPI will break.

Reply:
> @andybui01 can you expand upon this a bit? I'm actually investigating a problem where a VM is getting starved of execution time, and as far as I can tell, the userspace VCPU execution loops aren't the culprit.

If you are running multiple tasks on the same core that are woken up by notifications, perhaps #902 is related. To quickly test if this is the cause, enable signal fast paths and change the `SCHED_ENQUEUE` with `SCHED_ENQUEUE` at https://github.com/seL4/seL4/blob/0260fba168b74dd2504c9447f2fa8da1161ead87/src/fastpath/fastpath.c#L714

But try removing the whole if-check and calling schedule unconditionally first.

> I'm not really familiar with the x86 virtualization support but here's what I recall: currently the `VMCheckBoundNotification()` function that runs on the destination core may ask for a switch to a new thread. This new thread gets scheduled and activated in `handleInterruptEntry@syscall.c`. This can work because remote call IPIs operate with the acquired BKL of the calling core.
> 
> If we follow this PR so that `schedule()` checks that the _current_ core holds the BKL, this IPI will break.

Everyone takes the lock if IPI is not a remote call, but only x86 wants a schedule if the remote call is `IpiRemoteCall_VMCheckBoundNotification`.

I don't understand why x86 has a `IpiRemoteCall_VMCheckBoundNotification` IPI instead of just using the reschedule IPI.

Reply:
This is an x86 system, so the Notification fastpath isn't an option I can work with.

Reply:
Try removing the if-check and see if that solves your problem on x86.